design/vela-core/rollout-design.md

```diff
@@ -0,0 +1,269 @@
+}
+```
+
+## User Experience Workflow
+OAM rollout experience is different from flagger in a few different ways and here are the
+ implications on its impact on the user experience.
+- We assume that the resources it refers to are **_immutable_**. In contrast, flagger watches over a
+  target resource and reacts whenever the target's specification changes.
+    - The trait version of the controller refers to componentRevision.
+    - The application version of the controller refers to immutable application. 
+- The rollout logic **_works only once_** and stops after it reaches a terminal state. One can
+  still change the rollout plan in the middle of the rollout as long as it does not change the
+  pods that are already updated. 
+- The applicationDeployment controller only works with applications with 1 component for now.
+- Users should rely on the rollout CR to do the actual rollout which means they
+ shall set the `replicas` or `partition` field of the new resources to the starting value
+ indicated in the [detailed rollout plan design](#rollout-plan-work-with-different-type-of-workloads).
+- 
+    
+ 
+## Notable implementation level details
+Here are some high level implementation decisions
+
+### Rollout on different resources
+As we mentioned in the introduction section, we will implement two rollout controllers that work
+on different levels. At the end, they both emit an in-memory rollout plan object which includes
+references to the target and source kubernetes resources that the rollout planner will execute
+upon. For example, the applicationDeployment controller will get the component from the
+ application and extract the real workload from it before passing it to the rollout plan object.
+ 
+ With that said, two controllers operate differently to extract the real workload. Here are the
+  high level descriptions of how each works. 
+ 
+#### ApplicationDeployment takes control of the application during rolling out phase
+When an appDeployment is used to do application level rollout, **_one should apply the
+ appDeployment CR before the application itself_**. This is to try to make sure that the
+  appDeployment controll has full control of the new application from the beginning.
+ However, there is no kubernetes systematic way of acquiring a distributed lock for a controller.  
+- The appDeployment controller will make sure it marks itself as the owner of the
+ application and changes the appDeployment CR state to be initialized. The application controller
+  can be aware of appDeployment and have build-in logics in case.
+- The appDeployment controller can change the application fields. For example, it can remove all
+ the conflict traits, such as HPA, during the upgrade period. 
+- Upon a successful rollout, the appDeployment controller can remove all the traits in the old
+  application and leave no pods in the workload.
+- Upon a failed rollout, the condition is not determined, it could result in an unstable state.
```

Comment:
Why it's unstable? Even a failed rollout should make sure the replicas alive larger or equal then it should be or maxUnavaliable defined, right?

Reply:
since the appDeployment creates/remove traits, I am not sure if there is a way to guarantee the intermediate state with two application running is stable. It's not just the replica number.